### ISSUE TITLE
Cannot redeclare vomit function

### ISSUE THREAD

**Issue by tristanbes** [Interfacelab/ilab-media-tools#130]
Both in your standard edition and premium edition (we encounter the bug with the premium edition)

```json
  "autoload": {
    "psr-4": {
      "MediaCloud\\Plugin\\": "classes"
    },
    "files": [
      "classes/Utilities/Helpers.php"
    ]
  }
```

Problem is: you are manually autoloading the `classes/Utilities/Helpers.php` which is **already** autoloaded by the `"MediaCloud\\Plugin\\": "classes"` rule.

It result in a fatal error "Cannot redeclare function `vomit` ...`

You should release a patch version removing the `files` section in the composer.json

Correct version:
 ```json
  "autoload": {
    "psr-4": {
      "MediaCloud\\Plugin\\": "classes"
    }
  }

**Comment by jawngee** [comment]
psr-4 can't autoload functions, which is why it's setup like it is.  Removing the files section means Helpers never gets loaded.  Unless that's changed with Composer 2.x

Getting a "cannot redeclare" error means that Helpers is being loaded manually somewhere else, or there is another vomit function declared elsewhere.

We build client sites using bedrock (well we did until recently) so we're pretty experienced deploying with composer and with really heavy composer dependencies.  I've never experienced this error before.  Is there any way you can paste the entire stack.

**Comment by tristanbes** [comment]
Well, searching through the entire project (including vendors/plugins) only reports 1 function `vomit` that's why I had a hard time figuring the "fix".

![Capture d’écran 2021-02-09 à 17 28 43](https://user-images.githubusercontent.com/346010/107394552-4cef8300-6afc-11eb-92cc-6cdd69f12443.png)


Our stack is not ready yet for `composer v2` because of some `wp-cli` dependencies that are not released to support `composer 2` so I can confirm both the integrator and me got this issue running composer  `1.10.20`

Fullstack trace is:
```
Fatal error: Cannot redeclare MediaCloud\Plugin\Utilities\vomit() (previously declared in /srv/app/wordpress/restanques-provencales/web/app/plugins/ilab-media-tools-premium/classes/Utilities/Helpers.php:25) in /srv/app/wordpress/restanques-provencales/web/app/plugins/ilab-media-tools-premium/classes/Utilities/Helpers.php on line 25
```

Using the v4.1.14 of the premium version (`ilab-media-tools-premium`) on a private repositories of ours (that acts as a mirror) .  The error message is gone with the patch I suggested.

**Comment by jawngee** [comment]
You aren't really fixing it though is my point, somehow the file is being included twice, you are just removing one of the includes.

And the include has to be manual because composer's autoload code for files won't load a file twice.

I'm not saying it isn't something I'm doing, I'm saying your patch is not the right way to fix it.  And because we are 100% composer based for client work and I've never experienced this, something else is going on somewhere else.

Can you do a debug breakpoint on the include line in the composerRequire<UNIQUEID>() function of vendor/composer/autoload_real.php?

How are you using the premium version with composer?  Using Freemius's thing?

**Comment by tristanbes** [comment]
Yeah, it's a simple download from freemium, hosted on a private repositories because we were experiencing issues (can't remember which though).

Now that I have the specific version before my eyes, I can guess what's happening. The freemium download version includes the whole vendor file (with the composer folder, with its own autoload), resulting in loading everything twice I guess.

Since I have 2 `autoload_real.php` file, one from the vendor at the root of the project, and the other one in `plugins/ilab-media-tools-premium`

```
from root composer folder

string(97) "/srv/app/wordpress/restanques-provencales/vendor/composer/../symfony/polyfill-php80/bootstrap.php" string(103) "/srv/app/wordpress/restanques-provencales/vendor/composer/../symfony/deprecation-contracts/function.php" string(97) "/srv/app/wordpress/restanques-provencales/vendor/composer/../symfony/polyfill-ctype/bootstrap.php" string(100) "/srv/app/wordpress/restanques-provencales/vendor/composer/../symfony/polyfill-mbstring/bootstrap.php" string(97) "/srv/app/wordpress/restanques-provencales/vendor/composer/../symfony/polyfill-php73/bootstrap.php" string(105) "/srv/app/wordpress/restanques-provencales/vendor/composer/../symfony/polyfill-intl-grapheme/bootstrap.php" string(107) "/srv/app/wordpress/restanques-provencales/vendor/composer/../symfony/polyfill-intl-normalizer/bootstrap.php" string(99) "/srv/app/wordpress/restanques-provencales/vendor/composer/../symfony/string/Resources/functions.php" string(97) "/srv/app/wordpress/restanques-provencales/vendor/composer/../symfony/polyfill-php72/bootstrap.php" string(103) "/srv/app/wordpress/restanques-provencales/vendor/composer/../wp-cli/package-command/package-command.php" string(109) "/srv/app/wordpress/restanques-provencales/vendor/composer/../wp-cli/mustangostang-spyc/includes/functions.php" string(97) "/srv/app/wordpress/restanques-provencales/vendor/composer/../wp-cli/php-cli-tools/lib/cli/cli.php" string(99) "/srv/app/wordpress/restanques-provencales/vendor/composer/../wp-cli/cache-command/cache-command.php" string(101) "/srv/app/wordpress/restanques-provencales/vendor/composer/../wp-cli/entity-command/entity-command.php" string(107) "/srv/app/wordpress/restanques-provencales/vendor/composer/../wp-cli/extension-command/extension-command.php" string(93) "/srv/app/wordpress/restanques-provencales/vendor/composer/../wp-cli/db-command/db-command.php" string(105) "/srv/app/wordpress/restanques-provencales/vendor/composer/../wp-cli/scaffold-command/scaffold-command.php" string(101) "/srv/app/wordpress/restanques-provencales/vendor/composer/../wp-cli/export-command/export-command.php" string(105) "/srv/app/wordpress/restanques-provencales/vendor/composer/../wp-cli/checksum-command/checksum-command.php" string(101) "/srv/app/wordpress/restanques-provencales/vendor/composer/../wp-cli/import-command/import-command.php" string(99) "/srv/app/wordpress/restanques-provencales/vendor/composer/../wp-cli/media-command/media-command.php" string(117) "/srv/app/wordpress/restanques-provencales/vendor/composer/../wp-cli/wp-config-transformer/src/WPConfigTransformer.php" string(101) "/srv/app/wordpress/restanques-provencales/vendor/composer/../wp-cli/config-command/config-command.php" string(97) "/srv/app/wordpress/restanques-provencales/vendor/composer/../wp-cli/core-command/core-command.php" string(97) "/srv/app/wordpress/restanques-provencales/vendor/composer/../wp-cli/cron-command/cron-command.php" string(99) "/srv/app/wordpress/restanques-provencales/vendor/composer/../wp-cli/embed-command/embed-command.php" string(97) "/srv/app/wordpress/restanques-provencales/vendor/composer/../wp-cli/eval-command/eval-command.php" string(97) "/srv/app/wordpress/restanques-provencales/vendor/composer/../wp-cli/i18n-command/i18n-command.php" string(105) "/srv/app/wordpress/restanques-provencales/vendor/composer/../wp-cli/language-command/language-command.php" string(121) "/srv/app/wordpress/restanques-provencales/vendor/composer/../wp-cli/maintenance-mode-command/maintenance-mode-command.php" string(103) "/srv/app/wordpress/restanques-provencales/vendor/composer/../wp-cli/rewrite-command/rewrite-command.php" string(97) "/srv/app/wordpress/restanques-provencales/vendor/composer/../wp-cli/role-command/role-command.php" string(117) "/srv/app/wordpress/restanques-provencales/vendor/composer/../wp-cli/search-replace-command/search-replace-command.php" string(101) "/srv/app/wordpress/restanques-provencales/vendor/composer/../wp-cli/server-command/server-command.php" string(99) "/srv/app/wordpress/restanques-provencales/vendor/composer/../wp-cli/shell-command/shell-command.php" string(111) "/srv/app/wordpress/restanques-provencales/vendor/composer/../wp-cli/super-admin-command/super-admin-command.php" string(101) "/srv/app/wordpress/restanques-provencales/vendor/composer/../wp-cli/widget-command/widget-command.php" string(96) "/srv/app/wordpress/restanques-provencales/vendor/composer/../oscarotero/env/src/env_function.php" string(108) "/srv/app/wordpress/restanques-provencales/vendor/composer/../roots/wp-password-bcrypt/wp-password-bcrypt.php" string(134) "/srv/app/wordpress/restanques-provencales/vendor/composer/../../web/app/plugins/ilab-media-tools-premium/classes/Utilities/Helpers.php"

from plugin vendor folder which shouldnt be present

string(32) "85a81ab1badda9626343846e0ceed435" string(141) "/srv/app/wordpress/restanques-provencales/web/app/plugins/ilab-media-tools-premium/vendor/composer/../yuloh/bccomp-polyfill/src/functions.php" 

string(32) "a36a78a03571c7ba045ccfbb413ae093" string(134) "/srv/app/wordpress/restanques-provencales/web/app/plugins/ilab-media-tools-premium/vendor/composer/../../classes/Utilities/Helpers.php"```

**Comment by jawngee** [comment]
Ahh ok.  I will test this out this week then and maybe just do a class check in Helpers.php to return if the namespaced function already exists.  The vendor folder has to exist for freemius users who update from admin (and for wporg users).  It would be cool if Freemius excluded it for the version added via composer though.

**Comment by tristanbes** [comment]
maybe they do if you use their composer mirror repository ?
Don't worry though, i've removed it from our mirror.

**Comment by jawngee** [comment]
Fixed in 4.2.3.  Let me know if you have issues.
